frontends/src/terminal/screen_buf.rs

```diff
@@ -0,0 +1,227 @@
+}
+
+/// An iterator that produces instructions for updating a screen to match changes in a ScreenBuf.
+pub struct ScreenBufIter<'a> {
+    buf: &'a mut ScreenBuf,
+    current_style: Option<Style>,
+    cursor_pos: Option<Pos>,
+}
+
+impl ScreenBuf {
+    pub fn new() -> Self {
+        ScreenBuf { cells: Vec::new() }
+    }
+
+    /// Get `ScreenOp` instructions that describe all changes to the screen buffer since the last time this method was called.
+    pub fn drain_changes(&mut self) -> ScreenBufIter {
+        ScreenBufIter {
+            buf: self,
+            current_style: None,
+            cursor_pos: None,
+        }
+    }
+
+    pub fn resize(&mut self, size: Pos) {
+        self.cells = Vec::new();
+        let mut row = Vec::new();
+        row.resize_with(size.col as usize, Default::default);
+        self.cells.resize(size.row as usize, row);
+    }
+
+    pub fn clear(&mut self) {
+        let size = self.size();
+        self.resize(size);
+    }
+
+    pub fn write_str(&mut self, pos: Pos, s: &str, style: Style) -> Result<(), Error> {
+        let mut char_pos = pos;
+        for ch in s.chars() {
+            self.set_char_with_style(char_pos, ch, style)?;
+            char_pos = self.next_pos(char_pos).ok_or(Error::OutOfBounds)?;
+        }
+        Ok(())
+    }
+
+    pub fn shade_region(&mut self, region: Region, shade: Shade) -> Result<(), Error> {
+        let mut shade_rect = |rect: Rect| {
+            for r in rect.rows.0..rect.rows.1 {
+                for c in rect.cols.0..rect.cols.1 {
+                    let pos = Pos { row: r, col: c };
+                    self.set_shade(pos, shade)?;
+                }
+            }
+            Ok(())
+        };
+        shade_rect(region.body())?;
+        shade_rect(region.last_line())
+    }
+
+    pub fn set_style(&mut self, pos: Pos, style: Style) -> Result<(), Error> {
+        self.get_mut(pos)?.set_style(style);
+        Ok(())
+    }
+
+    fn set_char_with_style(&mut self, pos: Pos, ch: char, style: Style) -> Result<(), Error> {
+        let cell = self.get_mut(pos)?;
+        cell.set_char(ch);
+        cell.set_style(style);
+        Ok(())
+    }
+
+    fn set_shade(&mut self, pos: Pos, shade: Shade) -> Result<(), Error> {
+        self.get_mut(pos)?.set_shade(shade);
+        Ok(())
+    }
+
+    fn get(&self, pos: Pos) -> Result<CharCell, Error> {
+        self.cells
+            .get(pos.row as usize)
+            .and_then(|row| row.get(pos.col as usize))
+            .map(|cell| *cell)
+            .ok_or(Error::OutOfBounds)
+    }
+
+    fn get_mut(&mut self, pos: Pos) -> Result<&mut CharCell, Error> {
+        self.cells
+            .get_mut(pos.row as usize)
+            .and_then(|row| row.get_mut(pos.col as usize))
+            .ok_or(Error::OutOfBounds)
+    }
+
+    pub fn size(&self) -> Pos {
+        Pos {
+            row: self.cells.len() as u32,
+            col: self.cells.get(0).map(|row| row.len()).unwrap_or(0) as u16,
+        }
```

Comment:
TODO: handle weird edge case where `row=0` and `col>0`. To do so:
1. store `size` field
2. set it on construction
3. update it on `resize()`.
4. return it from this `size()` method.